### ISSUE TITLE
[ENH] CASTLE Algorithm

### ISSUE THREAD

**Issue by Manas-7854** [pgmpy/pgmpy#3140]
**Is your feature request related to a problem? Please describe.**
CASTLE is a regularization method that improves supervised learning by jointly learning a DAG and a predictive model over the data

**Describe the solution you'd like A clear and concise description of what you want to happen, ideally taking into consideration the existing toolbox design, classes and methods.**
### `_CASTLEModel(nn.Module)` (Internal)
PyTorch masked autoencoder for feature reconstruction, target prediction, and training.

- **`__init__(num_inputs, model_cfg: ModelConfig, reg_cfg: RegularizationConfig)`**: Initializes masked input layers, scalar output layers, grouped hyperparameters, and optimizer.
- **`forward(X)`**: Returns full reconstruction (`Out`) and target prediction (`out_0`).
- **`train(X_tensor)`**: Runs epochs with early stopping (batch updates, Lagrangian multiplier adjustments), logs to TensorBoard if enabled, and returns the thresholded adjacency matrix (`W_final`).
- **`get_W()`**: Computes adjacency matrix from current weights (L2 norm of column $j$ in sub-network $i$).

### Internal dataclasses
Grouped configuration objects used by `_CASTLEModel`.

- **`ModelConfig`**: `hidden_dim`, `batch_size`, `max_epochs`, `optimizer`, `seed`, `min_loss_improvement`, `early_stop_patience`, `tensorboard_log_dir`, `scaler`, `target_col`
- **`RegularizationConfig`**: `dag_weight`, `sparsity_weight`, `dag_penalty`, `edge_threshold`

### `CASTLE(_BaseCausalDiscovery)` (Public API)
Validates data, orchestrates training, and builds the `pgmpy.DAG`.

```python
CASTLE( 
    dag_weight=1.0,
    sparsity_weight=5.0,
    dag_penalty=1.0,
    optimizer=None,
    batch_size=32,
    hidden_dim=32,
    edge_threshold=0.3,
    target_col=None,
    max_epochs=200, 
    min_loss_improvement=1e-4,
    early_stop_patience=10,
    scaler=None,
    tensorboard_log_dir=None,
    seed=None,
)
```

Joint training objective being minimized:

$$\min_\Theta \frac{1}{N}\|Y - [f_\Theta(\tilde{X})]_{:,1}\|^2 + \lambda \underbrace{\left(L_N(f_\Theta) + (\text{tr}(e^{M \odot M}) - d - 1)^2 + \beta V_{\Theta_1}\right)}_{R_{\text{DAG}}}$$

- `dag_weight` (λ): Weight on the entire DAG regularization loss $R_{\text{DAG}}$.
- `sparsity_weight` (β): Group-lasso penalty on input-layer weights $V_{\Theta_1}$, promoting edge sparsity.
- `dag_penalty` (ρ): Initial penalty coefficient in the Augmented Lagrangian for the acyclicity constraint.
- `optimizer`: Any `torch.optim.Optimizer` instance. Defaults to `Adam(lr=1e-3)` if `None`. Note: `weight_decay` is not recommended when passing a custom optimizer as CASTLE has its own sparsity mechanism via `dag_weight` and `sparsity_weight`.
- `batch_size`: Number of samples per mini-batch.
- `hidden_dim` (h): Hidden layer width in each sub-network $f_k$.
- `edge_threshold`: Edges with weight below this value are zeroed in the final DAG.
- `target_col`: Target column name or index. Defaults to the first column when `None`.
- `max_epochs`: Maximum number of training epochs.
- `min_loss_improvement`: Minimum decrease in total loss per epoch to count as an improvement for early stopping.
- `early_stop_patience`: Number of consecutive epochs without improvement before stopping early.
- `scaler`: Optional scaler for input normalization (for example, `sklearn.preprocessing.StandardScaler`). It must implement `fit(X)` and `transform(X)`; `inverse_transform(X)` is optional for user-side de-scaling. If provided, it is fit on training data and stored as `scaler_`.
- `tensorboard_log_dir`: If set, logs training metrics via `SummaryWriter(log_dir=...)`. If `None`, TensorBoard logging is disabled.
- `seed`: Seed for reproducibility.

**Methods:**
- **`fit(X)`**: Trains the model and builds the DAG.

**Attributes set after `fit`:**
- `n_features_in_`: Number of input features seen during fit.
- `feature_names_in_`: Feature names from the input DataFrame.
- `causal_graph_`: Learned causal DAG as a `pgmpy.base.DAG`.
- `adjacency_matrix_`: Weighted adjacency matrix as a Pandas DataFrame.
- `model_`: Internal neural network instance used for training.
- `scaler_`: Fitted scaler used to normalize inputs during training (when provided).

---
## Usage

```python
import torch
from pgmpy.causal_discovery import CASTLE
import pandas as pd
import numpy as np
from sklearn.preprocessing import StandardScaler

df = pd.DataFrame(np.random.randn(100, 3), columns=['A', 'B', 'Target'])

# Default: Adam with lr=1e-3
model = CASTLE(max_epochs=200, edge_threshold=0.3, seed=42)

# Custom optimizer — user controls all optimizer params directly
opt = torch.optim.Adam(lr=1e-4, weight_decay=1e-5)
model = CASTLE(max_epochs=200, edge_threshold=0.3, optimizer=opt, seed=42)

# User-provided scaler and early stopping defaults
scaler = StandardScaler()
model = CASTLE(
    max_epochs=200,
    edge_threshold=0.3,
    target_col="Target",
    scaler=scaler,
    min_loss_improvement=1e-4,
    early_stop_patience=10,
    tensorboard_log_dir="runs/castle",
    seed=42,
)
model.fit(df)
dag = model.causal_graph_

# User-side scaling and inverse-scaling for predictions (example)
scaler_fitted = model.scaler_
X_scaled = scaler_fitted.transform(df.values)
y_pred_scaled = some_model(X_scaled)
y_pred = scaler_fitted.inverse_transform(y_pred_scaled)
```

View logs with: `tensorboard --logdir runs/castle`

**Additional context**
paper: https://arxiv.org/pdf/2009.13180
ref codebase: https://github.com/trentkyono/CASTLE

Description: related : #2931, pgmpy/enhancement_proposals#4

Note this issue is part of an ongoing mentorship project

**Comment by Manas-7854** [comment]
I have gone through the paper, here is a basic implementation plan:

I am unsure where should this be implemented since pgmpy currently doesn't have a separate directory for DL related algorithms
my suggestion would be to create a submodule inside `pgmpy/estimarors` for DL algorithms, this way we can isolate the DL dependencies i.e pytorch/tensorflow etc that these algorithms will require using optional dependency guard at import time.

For the tests (apart from the basic tests for the implmentations), we can:
- (similar to the implementation provided above), generate toy DAGs at three scales (4 nodes, 10 nodes, 20 nodes) and use them to generate synthetic data and see if the model is able to recover the toy DAGs
- Neural net generalization: compare test MSE across three variants on the same data — full CASTLE, CASTLE without DAG regularization (reg_lambda=0, reg_beta=0), and a plain MLP. Assert full CASTLE outperforms or matches both baselines on small n (200 samples) where the structural inductive bias matters most.
- assert that importing pgmpy and using classical estimators works correctly in an environment without PyTorch installed

Also I would suggest implementing this in pytorch to align with the broader DL ecosystem 

cc: @ankurankan

**Comment by holysheep123** [comment]
Hi **Manas-7854** 鈥?happy to help debug. To speed things up, could you share:

1. The exact error message (full text, not just the code)
2. What model/endpoint you're trying to use
3. Whether this was working before or is a fresh setup

鈥?holysheep.ai

**Comment by ankurankan** [comment]
This needs to go into `pgmpy/causal_discovery`, not `estimators`. `pytorch` is already a soft dependency. We also have a `pytorch` backend that switches to use pytorch for all computation instead of numpy.  You can also use `_safe_import` method to handle any optional dependencies (search for the method in the repo for examples).
